README.md
patiencebar
===========

Terminal progress bar compatible with multi-threading

Built by `Guillaume Schworer <https://github.com/ceyzeriat>`_ and collaborators (see
``corner.__contributors__`` for the most up to date list). Licensed under
the 2-clause BSD license (see ``LICENSE``).


Installation
------------

Just run

::

    pip install corner

to get the most recent stable version.


Usage
-----

The main entry point is the ``corner.corner`` function. You'll just use it
like this:

::

    import numpy as np
    import corner

    ndim, nsamples = 5, 10000
    samples = np.random.randn(ndim * nsamples).reshape([nsamples, ndim])
    figure = corner.corner(samples)
    figure.savefig("corner.png")

With some other tweaks (see `demo.py
<https://github.com/dfm/corner.py/blob/master/demo.py>`_) you can get
something that looks awesome like:

.. image:: https://raw.github.com/dfm/corner.py/master/corner.png

By default, data points are shown as grayscale points with contours.
Contours are shown at 0.5, 1, 1.5, and 2 sigma.

For more usage examples, take a look at `tests.py
<https://github.com/dfm/corner.py/blob/master/tests.py>`_.


Documentation
-------------

All the options are documented in the docstrings for the ``corner`` and
``hist2d`` functions. These can be viewed in a Python shell using:

::

    import corner
    print(corner.corner.__doc__)

or, in IPython using:

::

    import corner
    corner.corner?


A note about "sigmas"
+++++++++++++++++++++

We are regularly asked about the "sigma" levels in the 2D histograms. These
are not the 68%, *etc.* values that we're used to for 1D distributions. In two
dimensions, a Gaussian density is given by:

::

    pdf(r) = exp(-(r/s)^2/2) / (2*pi*s^2)

The integral under this density is:

::

    cdf(x) = Integral(r * exp(-(r/s)^2/2) / s^2, {r, 0, x})
           = 1 - exp(-(x/s)^2/2)

This means that within "1-sigma", the Gaussian contains ``1-exp(-0.5) ~ 0.393``
or 39.3% of the volume. Therefore the relevant 1-sigma levels for a 2D
histogram of samples is 39% not 68%. If you must use 68% of the mass, use the
``levels`` keyword argument.

The `"sigma-demo" notebook
<https://github.com/dfm/corner.py/blob/master/sigma-demo.ipynb>`_ visually
demonstrates the difference between these choices of levels.


Attribution
-----------

.. image:: https://zenodo.org/badge/4729/dfm/corner.py.svg
   :target: https://zenodo.org/badge/latestdoi/4729/dfm/corner.py

If you make use of this code, please `cite it
<https://zenodo.org/badge/latestdoi/4729/dfm/corner.py>`_.


License
-------

Copyright 2013, 2014 Dan Foreman-Mackey

corner.py is free software made available under the BSD License.
For details see the LICENSE file.



HISTORY.rst
1.0.0 (2016-05-03)
++++++++++++++++++

- Initial release.

example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

###############################################################################
#  
#  patiencebar - Terminal progress bar compatible with multi-threading
#  Copyright (C) 2016  Guillaume Schworer
#  
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#  
#  For any information, bug report, idea, donation, hug, beer, please contact
#    guillaume.schworer@obspm.fr
#
###############################################################################



import time
import numpy as np
import patiencebar as PB
from multiprocessing import Pool


# single-thread

n_calc = 34

pb = PB.Patiencebar(valmax=n_calc, barsize=50, title="Test bar")

for i in range(n_calc):
    time.sleep(np.random.random()*0.2)
    pb.update()


################################################################################
# multi-thread

def worker(params):
    pbm, num = params
    time.sleep(num*4)
    num = num**2
    pbm.update()
    return num


def worker2(params):
    i, pbm, num = params
    time.sleep(num*4)
    num = num**2
    pbm.update("Just got %ith element done: %f" % (i, num))
    return num


threads = 5
n_calc = 10
vector = np.random.random(n_calc)


pbm = PB.Patiencebarmulti(valmax=n_calc, barsize=30, title="Heavy calculation of the square of a vector")

pool = Pool(processes=threads)
result = pool.map_async(worker, ((pbm, n_calc) for i, n_calc in enumerate(vector)))
pool.close()
pool.join()
vector2 = np.asarray(result.get())


pbm.reset(bar=False)

pool = Pool(processes=threads)
result = pool.map_async(worker2, ((i, pbm, n_calc) for i, n_calc in enumerate(vector)))
pool.close()
pool.join()
vector2b = np.asarray(result.get())


patiencebar.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

###############################################################################
#  
#  patiencebar - Terminal progress bar compatible with multi-threading
#  Copyright (C) 2016  Guillaume Schworer
#  
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#  
#  For any information, bug report, idea, donation, hug, beer, please contact
#    guillaume.schworer@obspm.fr
#
###############################################################################


#from __future__ import print_function, absolute_import, unicode_literals

__all__ = ["Patiencebar", "Patiencebarmulti", "__version__", "__author__", "__copyright__", "__contributors__"]
__major__ = "1"
__minor__ = "0"
__version__ = "1.0.2"
__author__ = "Guillaume Schworer (guillaume.schworer@obspm.fr)"
__copyright__ = "Copyright 2016 Guillaume Schworer"
__contributors__ = [
    # Alphabetical by first name.
]

import os as _os
from sys import stdout as _stdout
from threading import Thread as _Thread
from time import sleep as _sleep
from multiprocessing import Manager as _Manager


class Patiencebar(object):
    """
    Provides a terminal-friendly single-thread progress bar

    Args:
      * valmax (float): the finish value of the progress bar. Default is 100.
      * barsize (int >0): the size of the bar in the opened terminal. If ``None``, the bar will automatically fit the width of the window.
      * title (str): the title, printed one line above the progress bar
      * bar (bool): whether the bar should be displayed or not. If ``False``, only the text given at each :func:`update` will be printed
      * up_every (int [0-100]): if ``bar`` is ``True``, the progress bar will be updated every ``up_every`` percent of progress. Setting ``up_every`` = 0 updates the progress bar at each :func:`update`

    >>> import patiencebar as PB
    >>> n_calc = 34
    >>> pb = PB.Patiencebar(valmax=n_calc, barsize=50, title="Test bar")
    >>> for i in range(n_calc):
    >>>     do_stuff()
    >>>     pb.update()
    """
    def __init__(self, valmax=100, barsize=None, title=None, bar=True, up_every=2):
        self.reset(valmax=valmax, barsize=barsize, title=title, bar=bar, up_every=up_every)

    @property
    def valmax(self):
        return self._valmax
    @valmax.setter
    def valmax(self, value):
        raise AttributeError("Read-only")

    @property
    def barsize(self):
        return self._barsize
    @barsize.setter
    def barsize(self, value):
        raise AttributeError("Read-only")

    @property
    def title(self):
        return self._title
    @title.setter
    def title(self, value):
        raise AttributeError("Read-only")

    @property
    def running(self):
        return self._running
    @running.setter
    def running(self, value):
        raise AttributeError("Read-only")

    @property
    def bar(self):
        return self._bar
    @bar.setter
    def bar(self, value):
        raise AttributeError("Read-only")

    @property
    def up_every(self):
        return self._up_every
    @up_every.setter
    def up_every(self, value):
        raise AttributeError("Read-only")


    def reset(self, valmax=None, barsize=None, title=None, bar=None, up_every=None):
        """
        Resets the progress bar with initialization values, unless new values are given

        Args:
          * valmax (float): the finish value of the progress bar. Default is 100.
          * barsize (int >0): the size of the bar in the opened terminal. If ``None``, the bar will automatically fit the width of the window.
          * title (str): the title, printed one line above the progress bar.
          * bar (bool): whether the bar should be displayed or not. If ``False``, only the text given at each :func:`update` will be printed.
          * up_every (int [0-100]): if ``bar`` is ``True``, the progress bar will be updated every ``up_every`` percent of progress. Setting ``up_every`` = 0 updates the progress bar at each :func:`update`.

        >>> import patiencebar as PB
        >>> n_calc = 34
        >>> pb = PB.Patiencebar(valmax=n_calc, barsize=50, title="Test bar")
        >>> for i in range(n_calc):
        >>>     do_stuff()
        >>>     pb.update()
        >>> pb.reset(title="Second trial", barsize=70)
        >>> for i in range(n_calc):
        >>>     do_stuff()
        >>>     pb.update()
        """
        self.win_width = _get_terminal_size()[1]
        self.step = 0
        self._nextup = 0
        self._title_written = False
        self._running = False
        if valmax is not None:
            self._valmax = float(valmax)
        else:
            self._valmax = float(getattr(self, 'valmax', 100))
        if barsize is not None:
            self._barsize = int(barsize)
        else:
            self._barsize = int(getattr(self, 'barsize', self.win_width-8))
        if title is not None:
            self._title = str(title)
        elif getattr(self, 'title', None) is not None:
            self._title = str(getattr(self, 'title'))
        else:
            self._title = None
        if bar is not None:
            self._bar = bool(bar)
        else:
            self._bar = bool(getattr(self, 'bar', True))
        if up_every is not None:
            self._up_every = int(min(100,max(0,up_every)))
        else:
            self._up_every = int(getattr(self, 'up_every', 2))

    
    def update(self, step=None):
        """
        Updates the progress bar to a newer value

        Args:
          * step (None): adds 1 to the progress of the bar, where ``valmax`` is the finish value.
          * step (float): sets the progress of the bar to the ``step`` value, where ``valmax`` is the finish value.
          * step (str): displays ``step`` on a new line. For this to work, ``bar`` must be ``False`` (no progress bar displayed) otherwise the update instruction is ignored.
        """
        self._doupdate(step=step)
        
    def _doupdate(self, step=None):
        # initialization
        if not self._title_written:
            if self.title is not None: print(self.title)
            self._title_written = True
            self._running = True # in case it was not yet done
        if self.bar: # if display progressbar
            if step is None: # no step given
                self.step += 1
            else: # step given              
                try: # make sure we got some number
                    self.step = min(float(step), self.valmax)
                except: # if not.. WTF.. don't change anything
                    return
            # render bar
            perc = int(self.step*100./self.valmax)
            if perc < self._nextup: return
            self._nextup = min(100, perc+self._up_every)
            bar = int(perc/100.*self.barsize)
            spacing = (self.win_width-self.barsize-8)/2.
            out = '\r%s[%s%s] %3d %%%s' % (' ' * int(spacing), '=' * bar, ' ' * int(self.barsize - bar), perc, ' ' * int(spacing+0.5))
            _stdout.write(out)
            _stdout.flush()
        else: # if no bar display
            if step is None: step = 'tic'
            print(str(step))
            self.step += 1
        if self.step >= self.valmax: # finished
            print("\r")
            self._running = False            


class Patiencebarmulti(Patiencebar):
    """
    Provides a terminal-friendly multi-thread progress bar

    Args:
      * valmax (float): the finish value of the progress bar. Default is 100.
      * barsize (int >0): the size of the bar in the opened terminal. If ``None``, the bar will automatically fit the width of the window.
      * title (str): the title, printed one line above the progress bar
      * bar (bool): whether the bar should be displayed or not. If ``False``, only the text given at each :func:`update` will be printed
      * up_every (int [0-100]): if ``bar`` is ``True``, the progress bar will be updated every ``up_every`` percent of progress. Setting ``up_every`` = 0 updates the progress bar at each :func:`update`

    >>> import patiencebar as PB
    >>> from threading import Thread
    >>> n_calc = 34
    >>> 
    >>> def worker(pbm, otherarg, anotherarg):
    >>>     do_stuff(otherarg, anotherarg)
    >>>     pbm.update()
    >>> 
    >>> pbm = PB.Patiencebarmulti(n_calc, 50, "Test bar")
    >>> for i in range(n_calc):
    >>>     ttt = Thread(target=worker, args=(pbm, otherarg, anotherarg))
    >>>     ttt.daemon = True
    >>>     ttt.start()
    """
    def __init__(self, valmax=100, barsize=None, title=None, bar=True, up_every=2):
        self._q = _Manager().Queue(maxsize=0)
        self.reset(valmax=valmax, barsize=barsize, title=title, bar=bar, up_every=up_every)

    def reset(self, valmax=None, barsize=None, title=None, bar=None, up_every=None):
        """
        Resets the progress bar with initialization values, unless new values are given

        Args:
          * valmax (float): the finish value of the progress bar. Default is 100.
          * barsize (int >0): the size of the bar in the opened terminal. If ``None``, the bar will automatically fit the width of the window.
          * title (str): the title, printed one line above the progress bar.
          * bar (bool): whether the bar should be displayed or not. If ``False``, only the text given at each :func:`update` will be printed.
          * up_every (int [0-100]): if ``bar`` is ``True``, the progress bar will be updated every ``up_every`` percent of progress. Setting ``up_every`` = 0 updates the progress bar at each :func:`update`.
        """
        super(Patiencebarmulti, self).reset(valmax=valmax, barsize=barsize, title=title, bar=bar, up_every=up_every)
        self._running = True
        checker = _Thread(target=self._check)
        checker.setDaemon(True)
        checker.start()

    def stop(self):
        """
        Stops the progress bar, in case it didn't stop naturally
        """
        self._running = False

    def _check(self):
        while self.running:
            step = self._q.get()
            self._doupdate(step=step)
            self._q.task_done()
            _sleep(0.1)
        self._running = False

    def update(self, step=None):
        """
        Updates the progress bar to a newer value

        Args:
          * step (None): adds 1 to the progress of the bar, where ``valmax`` is the finish value.
          * step (float): sets the progress of the bar to the ``step`` value, where ``valmax`` is the finish value.
          * step (str): displays ``step`` on a new line. For this to work, ``bar`` must be ``False`` (no progress bar displayed) otherwise the update instruction is ignored.
        """
        self._q.put(step)

def _get_terminal_size():
    def ioctl_GWINSZ(fd):
        try:
            import fcntl, termi_os, struct
            cr = struct.unpack('hh', fcntl.ioctl(fd, termi_os.TIOCGWINSZ,'1234'))
        except:
            return
        return cr
    cr = ioctl_GWINSZ(0) or ioctl_GWINSZ(1) or ioctl_GWINSZ(2)
    if not cr:
        try:
            fd = _os.open(_os.ctermid(), _os.O_RDONLY)
            cr = ioctl_GWINSZ(fd)
            _os.cl_ose(fd)
        except:
            pass
    if not cr: cr = (_os.environ.get('LINES', 25), _os.environ.get('COLUMNS', 80))
    return map(int, cr)
